ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintExportType.java

```diff
@@ -0,0 +1,164 @@
+          String defaultValue = defaultConfig.getPropertyValue(configType, propertyName);
+          if (include(propertyValue, defaultValue))  {
+            LOG.debug("Including {}/{} in exported blueprint, as default value and actual value differ:\n{}\nvs\n{}", configType, propertyName, defaultValue, propertyValue);
+          } else {
+            LOG.debug("Omitting {}/{} from exported blueprint, as it has the default value of {}", configType, propertyName, propertyValue);
+            actualConfig.removeProperty(configType, propertyName);
+          }
+        }
+        if (properties.isEmpty()) {
+          actualConfig.getProperties().remove(configType);
+        }
+      }
+
+      for (Map.Entry<String, Map<String, Map<String, String>>> configTypeEntry : ImmutableSet.copyOf(actualConfig.getAttributes().entrySet())) {
+        String configType = configTypeEntry.getKey();
+        Map<String, Map<String, String>> attributes = configTypeEntry.getValue();
+        for (Map.Entry<String, Map<String, String>> attributeEntry : ImmutableSet.copyOf(attributes.entrySet())) {
+          String attributeName = attributeEntry.getKey();
+          Map<String, String> properties = attributeEntry.getValue();
+          for (Map.Entry<String, String> propertyEntry : ImmutableSet.copyOf(properties.entrySet())) {
+            String propertyName = propertyEntry.getKey();
+            String attributeValue = propertyEntry.getValue();
+            String defaultValue = defaultConfig.getAttributeValue(configType, propertyName, attributeName);
+            if (include(attributeValue, defaultValue))  {
+              LOG.debug("Including {}/{}/{} in exported blueprint, as default value and actual value differ:\n{}\nvs\n{}", configType, attributeName, propertyName, defaultValue, attributeValue);
+            } else {
+              LOG.debug("Omitting {}/{}/{} from exported blueprint, as it has the default value of {}", configType, attributeName, propertyName, attributeValue);
+              properties.remove(propertyName);
+            }
+          }
+          if (properties.isEmpty()) {
+            attributes.remove(attributeName);
+          }
+        }
+        if (attributes.isEmpty()) {
+          actualConfig.getAttributes().remove(configType);
+        }
+      }
+
+      return actualConfig;
+    }
+
+    @Override
+    public boolean include(String value, String defaultValue) {
+      return value != null && (
+        defaultValue == null ||
+          !Objects.equals(StringUtils.trim(defaultValue), StringUtils.trim(value))
+      );
+    }
+
+    @Override
+    public boolean include(Collection<?> collection) {
+      return collection != null && !collection.isEmpty();
+    }
+
+    @Override
+    public boolean include(Map<?, ?> map) {
+      return map != null && !map.isEmpty();
+    }
+  },
+  ;
+
+  public abstract Configuration filter(Configuration actualConfig, Configuration defaultConfig);
```

Comment:
The filter method is called twice from _ClusterBlueprintRenderer::createBlueprintResource_

1. via createClusterTopology -> ExportBlueprintRequest::new -> ExportBlueprintRequest::createConfiguration
1. via BlueprintConfigurationProcessor::doUpdateForBlueprintExport

Reply:
ExportBlueprintRequest seemed the  place to handle it.  But some conversion logic is only possible in BlueprintConfigurationProcessor, so I added it there, too.  It turns out that ExportBlueprintRequest no longer needs to filter, so I removed it from there.